File: magmac/src/java/magma/api/contain/stream/Stream.java
package magma.api.contain.stream;

import magma.api.Tuple;
import magma.api.contain.collect.Collector;
import magma.api.option.Option;
import magma.api.result.Result;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

public interface Stream<T> extends Head<T> {
    <R> Stream<R> map(Function<T, R> mapper);

    <C> C collect(Collector<T, C> collector);

    <C> C foldLeft(C current, BiFunction<C, T, C> folder);

    Option<T> head();

    <R, E> Result<R, E> foldLeftToResult(R initial, BiFunction<R, T, Result<R, E>> mapper);

    boolean anyMatch(Predicate<T> predicate);

    Stream<T> filter(Predicate<T> filter);

    <R> Stream<R> flatMap(Function<T, Head<R>> mapper);

    <R> Stream<Tuple<T, R>> extend(Function<T, R> mapper);

    default <R> R into(Function<Stream<T>, R> mapper) {
        return mapper.apply(this);
    }
}

File: src/clang/magma/collect/stream/Stream.c
#include "Stream.h"
magma.collect.stream.R foldWithInitial(magma.collect.stream.R initial, magma.collect.stream.R(*folder)(magma.collect.stream.R, T));
magma.collect.stream.Stream<magma.collect.stream.R> map(magma.collect.stream.R(*mapper)(T));
magma.collect.stream.C collect(magma.collect.stream.Collector<T, magma.collect.stream.C> collector);
magma.collect.stream.Stream<magma.collect.stream.R> flatMap(magma.collect.stream.Stream<magma.collect.stream.R>(*mapper)(T));
magma.result.Result<magma.collect.stream.R, magma.collect.stream.X> foldToResult(magma.collect.stream.R initial, magma.result.Result<magma.collect.stream.R, magma.collect.stream.X>(*folder)(magma.collect.stream.R, T));
magma.option.Option<magma.collect.stream.R> foldMapping(magma.collect.stream.R(*mapper)(T), magma.collect.stream.R(*folder)(magma.collect.stream.R, T));
magma.collect.stream.Stream<T> filter(int(*predicate)(T));
magma.collect.stream.Stream<T> concat(magma.collect.stream.Stream<T> other);
magma.option.Option<T> next();
magma.collect.stream.boolean anyMatch(int(*predicate)(T));

